### ISSUE TITLE
Evaluation API

### ISSUE THREAD

**Issue by ananyahjha93** [mosaicml/composer#1429]
## 🚀 Feature Request

Adding more ways to use the current eval API.

## Motivation

The eval API feels a bit restrictive and requires us to write duplicate code. For large eval harnesses, or places where only eval is required (nearest neighbor augmented LMs, a whole bunch of large pre-trained LMs etc.); passing ``eval_dataloader`` to trainer is not sufficient for calling ``trainer.eval()``.

If ``eval_dataloader`` is passed to the trainer, I am assuming the Composer model will already have ``validate`` and ``metric`` functions.

```
trainer = Trainer(
  eval_dataloader=...,
)

# pass dataloader and metric again
# can it reuse metric defined in composer model?
trainer.eval(dataloder, metric)
```

Simply calling ``trainer.eval()`` without the need for passing additional ``dataloader``, ``metric`` will also be useful for the case where the trainer already has been passed multiple evaluators (this is specially a really nice API design). 

```
eval1 = Evaluator(...)
eval2 = Evaluator(...)

trainer = Trainer(
eval_dataloader=[eval1, eval2],
)

trainer.eval()
```

## Additional context

Initial discussion on slack in composer channel.

**Comment by ananyahjha93** [comment]
@hanlint happy to continue the discussion here and if this change makes sense for the API.

**Comment by hanlint** [comment]
Thanks @ananyahjha93, I'm thinking we can revamp the `eval` API to mirror the `trainer` API, so that users can either do:
```
trainer = Trainer(eval_dataloader=[eval1, eval2, ..])
trainer.eval()
```
or
```
trainer = Trainer(...)
trainer.eval(
    eval_dataloader=[eval1, eval2]
)
```

With that design, the eval API would be something like:
```
def eval(self, 
    eval_dataloader: Union[DataLoader, Evaluator], 
    metrics: Optional[Metric] = None
):
```
where user can either provide an `dataloader` + `metrics`, or just provide evaluators.

Would that make sense?

**Comment by ananyahjha93** [comment]
@hanlint yes, that definitely makes the API uniform

**Comment by hanlint** [comment]
OK great, we'll implement this in the next release (v0.10), while maintaining some backwards compatibilty with the current API.
